agent/lib/kontena/observable.rb

```diff
@@ -1,75 +1,201 @@
 module Kontena
-  # An Actor that has some value
-  # The value does not yet exist when initialized, it is nil
-  # Once the value is first updated, then other Actors will be able to observe it
-  # When the value later updated, other Actors will also observe those changes
-  module Observable
+  # The observable value is nil when initialized, and Observers will block waiting for it to become ready.
+  # Once the observable is first updated, then Observers will unblock and return/yield the updated value.
+  # When the observable is later updated, then Observers will return/yield the updated value.
+  # If the observable is crashed, then any Observers will also raise.
+  #
+  # TODO: are you allowed to reset an observable after crashing it, allowing observers to restart and re-observe?
+  #
+  # @attr observers [Hash{Kontena::Observer => Boolean}] stored value is the persistent? flag
+  class Observable
+    require_relative './observable/registry'
+
+    # @return [Celluloid::Proxy::Cell<Kontena::Observable::Registry>] system registry actor
+    def self.registry
+      Celluloid::Actor[:observable_registry] || fail(Celluloid::DeadActorError, "Observable registry actor not running")
+    end
+
     include Kontena::Logging
 
-    def observable_value
-      @observable_value
+    attr_reader :logging_prefix # customize Kontena::Logging#logging_prefix by instance
+
+    class Message
+      attr_reader :observer, :observable, :value
+
+      # @param observer [Kontena::Observer]
+      # @param observable [Kontena::Observable]
+      # @param value [Object, nil, Exception]
+      def initialize(observer, observable, value)
+        @observer = observer
+        @observable = observable
+        @value = value
+      end
+    end
+
+    # mixin for Celluloid actor classes
+    module Helper
+      # Create a new Observable using the including class name as the subject.
+      # Register the Observable with the Kontena::Observable::Registry.
+      # Links to the registry to crash the Observable if the owning actor crashes.
+      #
+      # @return [Kontena::Observable]
+      def observable
+        @observable ||= Kontena::Observable.new(self.class.name).tap do |observable|
```

Comment:
Turns out that this `@observable ||= ... tap` is racy even within a single actor, probably because of the sync `observable_registry.register` call inside the `tap` block:

```
I, [2017-08-30T11:50:54.140808 #10290]  INFO -- Kontena::Observable::Registry: initialized
D, [2017-08-30T11:50:54.243572 #10290] DEBUG -- Kontena::Observable::Registry: register Kontena::Observable<TestObservableActor>: TestObservableActor
D, [2017-08-30T11:50:54.244124 #10290] DEBUG -- Kontena::Observable::Registry: register Kontena::Observable<TestObservableActor>: TestObservableActor
D, [2017-08-30T11:50:54.244522 #10290] DEBUG -- Kontena::Observer<TestObserverActor>: observe Kontena::Observable<TestObservableActor>...
D, [2017-08-30T11:50:54.244692 #10290] DEBUG -- Kontena::Observer<TestObserverActor>: wait: Kontena::Observable<TestObservableActor>?
D, [2017-08-30T11:50:54.244744 #10290] DEBUG -- Kontena::Observer<TestObserverActor>: receive timeout=...
D, [2017-08-30T11:50:54.244864 #10290] DEBUG -- Kontena::Observable<TestObservableActor>: update: 1
D, [2017-08-30T11:50:54.245628 #10290] DEBUG -- Kontena::Observer<TestObserverActor>: observe Kontena::Observable<TestObservableActor>...
D, [2017-08-30T11:50:54.245697 #10290] DEBUG -- Kontena::Observer<TestObserverActor>: wait: Kontena::Observable<TestObservableActor>?
D, [2017-08-30T11:50:54.245738 #10290] DEBUG -- Kontena::Observer<TestObserverActor>: receive timeout=...
D, [2017-08-30T11:50:54.246087 #10290] DEBUG -- Kontena::Observable<TestObservableActor>: update: 2
D, [2017-08-30T11:50:54.246764 #10290] DEBUG -- Kontena::Observer<TestObserverActor>: observe Kontena::Observable<TestObservableActor>...
D, [2017-08-30T11:50:54.246873 #10290] DEBUG -- Kontena::Observer<TestObserverActor>: wait: Kontena::Observable<TestObservableActor>?
D, [2017-08-30T11:50:54.246918 #10290] DEBUG -- Kontena::Observer<TestObserverActor>: receive timeout=...
```

The actor ends up with two different `@observable`s, and updates one while the observers try and observe the other one:

```
Failures:

  1) Kontena::Observable registered by an Actor handles concurrent observers
     Failure/Error: expect(observer_actors.map{|actor| actor.observed_values.last}).to eq [update_count] * observer_count
     
       expected: [10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10]
            got: [nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil]
```